### DISCUSSION TITLE
Stm32F7 HID host example

### DISCUSSION THREAD

**Discussion by arkadiuszmakarenko** [hathach/tinyusb#1158] [Q&A] [answered]
Does anyone have any example how to implement callbacks functions for stm32F7 (but any  F family would do) host HID?

**Comment by hathach** [comment, marked as answer]
currently there is no host driver for stm32 just yet, only rp2040 and nxp lpc chip are supported.

**Reply by arkadiuszmakarenko** [reply]
Do you think someone attempted this before?

I will try to use ST USB Low Level Drivers, and let see how far I will go (probably not far, but I will try)

**Reply by hathach** [reply]
Implementing the host controller driver (hcd) for stm32 will probably be quicker than using st low level drivers.

**Reply by arkadiuszmakarenko** [reply]
It is beyond my knowledge, unfortunately.

Definitely will be looking for updates of this project :)

**Reply by dgendreau** [reply]
I am an experienced embedded fw engineer but just starting on TinyUsb this week. I would like try to add host mode support for stm32. @hathach how can I help / where should I start on this? Is there a Discord where TUSB devs can chat?

**Reply by hathach** [reply]
thank you for your willingness. Unfortunately, there isn't much documentation for porting an host driver. Though it is similar to porting an device driver 
- https://github.com/hathach/tinyusb/blob/master/docs/contributing/porting.rst
- Basically implementing [hcd.h API](https://github.com/hathach/tinyusb/blob/master/src/host/hcd.h)

There could be some out of date docs/code (not having enough time to clean it up). You could take a look at this PR as referenece
- https://github.com/hathach/tinyusb/blob/master/src/host/hcd.h

> Is there a Discord where TUSB devs can chat?

We could use this github discussion for discussing, since my timezone is too off for most people and  I am not very active with discord as well. Just put up your questions here when you need help, I and others will try to help when possible.

PS: I also have an plan to add HCD for dwc2 as well in the not far away future, since there is lots of mcu use this driver.

**Reply by dgendreau** [reply]
Thank you. I will start reviewing these and post questions as I have them.

When I previously looked into ST's host drivers and found a few things lacking. They have a 1:1 mapping between pipes, connected endpoints and the 8-16 available hardware USB host peripheral channels. That is fine if you only plan to connect one device, but when you bring hubs into the picture you will quickly run out of channels. I believe we will need to add a mapping layer that queues I/O packets and dynamically re-allocates these host channels as needed.

We might also use this onboarding process as an opportunity to help improce and document the steps necessary for porting a host driver.

**Reply by Ryzee119** [reply]
I have a personal project using stm32f7xx and have implemented TinyUSB host driver using cube HAL drivers if curious as a starting point.

https://gist.github.com/Ryzee119/5a1b74e98449fdf5ed24b3f361ff4e2c

Far from PR ready but it atleast works with HID, USB Hubs and a few custom drivers I have written. Dont see why it wouldnt work with bulk,

You just need to enable to USB clocks, USB GPIO and link the hcd_int_handler to the OTG IRQ. This is from my F750-dk disco board
```
HCD_HandleTypeDef hhcd_USB_OTG_FS;

//Forward the OTG_FS interrupt to the tinyusb handler
extern "C" void hcd_int_handler(uint8_t rhport);
extern "C" void OTG_FS_IRQHandler(void)
{
    hcd_int_handler(0);
}

void usbh_dev_init()
{
    //Enable clocks
    __HAL_RCC_GPIOA_CLK_ENABLE();
    __HAL_RCC_GPIOD_CLK_ENABLE();

    GPIO_InitTypeDef GPIO_InitStruct;

    //Turn on 5V output
    GPIO_InitStruct.Pin = OTG_FS_PowerSwitchOn_Pin;
    GPIO_InitStruct.Mode = GPIO_MODE_OUTPUT_PP;
    GPIO_InitStruct.Pull = GPIO_NOPULL;
    GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_LOW;
    HAL_GPIO_Init(OTG_FS_PowerSwitchOn_GPIO_Port, &GPIO_InitStruct);
    HAL_GPIO_WritePin(OTG_FS_PowerSwitchOn_GPIO_Port, OTG_FS_PowerSwitchOn_Pin, GPIO_PIN_RESET);

    //Setup USB data pins
    GPIO_InitStruct.Pin = GPIO_PIN_11 | GPIO_PIN_12;
    GPIO_InitStruct.Mode = GPIO_MODE_AF_PP;
    GPIO_InitStruct.Pull = GPIO_NOPULL;
    GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_VERY_HIGH;
    GPIO_InitStruct.Alternate = GPIO_AF10_OTG_FS;
    HAL_GPIO_Init(GPIOA, &GPIO_InitStruct);

    __HAL_RCC_USB_OTG_FS_CLK_ENABLE();
    HAL_NVIC_SetPriority(OTG_FS_IRQn, 6, 0);
    HAL_NVIC_EnableIRQ(OTG_FS_IRQn);
}
```

**Reply by Ryzee119** [reply]
>When I previously looked into ST's host drivers and found a few things lacking. They have a 1:1 mapping between pipes, connected endpoints and the 8-16 available hardware USB host peripheral channels. That is fine if you only plan to connect one device, but when you bring hubs into the picture you will quickly run out of channels. I believe we will need to add a mapping layer that queues I/O packets and dynamically re-allocates these host channels as needed.

Yes this is right and I found this quite limiting. A couple usb hubs and devices and I quickly ran out of channels. I created a interrupt transfer queue and activated the interrupts channels at bIntervals if a channel was available. The channel is then released on a NAK so something else could use it until the next bInterval. Ive successfuly used 4 HID devices and 2 hubs at once (~10 interrupt endpoints)

The way TinyUSB is written (shared control buffer), I use one channel for all control transfers which works well for now.

Bulk and Iso have their own dedicated channel.

**Reply by Ryzee119** [reply]
Another caveat I encountered fot STM32F7 is this: (Ref OTG_HCTSIZx in F7xx ref manual)
> XFRSIZ[18:0]: Transfer size 
For an IN, this field is the buffer size that the application has reserved for the transfer. The
application is expected to program this field as an integer multiple of the maximum packet
size for IN transactions (periodic and non-periodic)

TinyUSB does not always supply a buffer that meets this criteria and will just overflow if used (usb hubs are a culprit!)

**Comment by AndrewCapon** [comment]
Hi @Ryzee119 

Did you ever find a perfect way of handling the dreaded `URB_NOTREADY`?

Thanks

Andy

**Comment by hathach** [comment]
update: hcd driver for dwc2 is implemented
